### PR TITLE
Switch to mocha

### LINKED ISSUE #22: Switch to Mocha
Reopening this, since it looks like one of our other issues got closed. I looked into test frameworks a bit more.

 - ava seems awesome, but I'm concerned with its concurrent nature. I'd imagine our tests involve modifying state on disk, so concurrently running tests each modifying disk state could be a big issue. If ava has a good workaround for this (other than the `.serial` method for every test case, which sounds like a pain), feel free to correct me.
 - mocha is pretty popular, and runs things concurrently. That's a bit slower than ava since it's not in parallel, but should be more than fast enough.
 - [rocha](https://github.com/bahmutov/rocha) seems pretty promising. This is a wrapper on top of mocha (so it's still running tests concurrently), but it runs them in a random order (this makes sure we don't have a dependency between different test cases). If we get a random order that fails, rocha will keep that order until we get the tests to pass (so it's still easy to debug issues). If we're starting from scratch, it should be easy to write tests that don't depend on order.

I don't care too much which framework we use, as long as it works and ties into coverage tools. I'm mostly posting this as a reminder that we need some sort of unit tests.

### PR DESCRIPTION
**Pull request by levithomason**
Fixes #22

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,11 +15,11 @@
...
     "lint:fix": "npm run lint --silent -- --fix",
-    "lint:watch": "watch 'npm run lint --silent' src",
+    "lint:watch": "watch 'npm run lint --silent' src test",
     "prepublish": "npm run build",
-    "pretest": "npm run lint",
-    "test": "ava",
-    "test:watch": "watch 'npm test --silent' src test"
+    "posttest": "npm run lint --silent",
+    "test": "mocha",
+    "test:watch": "concurrently -rk 'npm run test --silent -- -w' 'npm run lint:watch'"
```
Comment (levithomason) [inline package.json]: Moved lint to `posttest` so the lint results show after the tests run.  Helps keep linter happy while testing.

File: test/setup.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import sinonChai from 'sinon-chai';
+
+//
+// Sinon
+//
+global.sandbox = sinon.sandbox.create();
+
+afterEach(() => {
+  global.sandbox.restore();
+});
```
Comment (levithomason) [inline test/setup.js]: Global sandbox for spies/stubs/mocks that is auto restored after every test.

File: test/setup.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+//
+// Chai
+//
+global.expect = chai.expect;
+chai.should();
+chai.use(dirtyChai);
+chai.use(sinonChai);
```
Comment (levithomason) [inline test/setup.js]: [dirty-chai](https://github.com/prodatakey/dirty-chai) allows for function form for terminating assertion properties:

```js
should.be.true();
// vs
should.be.true;
```

This also keeps the linter happy and just makes more sense.

[sinon-chai](https://github.com/domenic/sinon-chai) adds chai assertions for sinon mock/spy/stubs.  This makes the assertion errors much more helpful as well.

```js
spy.called.should.be.ok();
// vs
spy.should.have.been.called();
```
>throws something like: "expected spy to been called at least once, but was not called."

File: test/specs/cli.js [hunk 4]
```diff
@@ -1,9 +1,5 @@
-import test from 'ava';
-import sinon from 'sinon';
```
Comment (levithomason) [inline test/specs/cli.js]: With the globals we have now, you don't need to import testing rig modules.  Just the stuff you are testing.

File: test/setup.js (hunk 3 above)
Comment (nfischer) [inline test/setup.js]: SGTM

**Comment by levithomason** [comment]
@ariporad any ideas why travis / appveyor are not running on this PR?

**Comment by nfischer** [comment]
@levithomason 

> any ideas why travis / appveyor are not running on this PR?

I believe it's configured to only run if `.travis.yml` is present, which it isn't on this branch. I will verify that this works for node v4 and node v5 for my system.

**Comment by levithomason** [comment]
Oh, right, it isn't merged yet :)

**Comment by nfischer** [comment]
@levithomason I got this error on node v4:

```
Error: Cannot find module 'babel-register'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:289:25)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at /home/nate/PR_31/node_modules/shx/node_modules/mocha/bin/_mocha:302:3
    at Array.forEach (native)
    ...
```

**Comment by levithomason** [comment]
Did you try installing deps again, or blowing away and reinstalling?  NPM has been having issues with that.  `babel-register` is needed by mocha to compile, however, it is included with `babel` which is already a dep.  This should be working.

**Comment by nfischer** [comment]
It looks like `babel-register` and `babel-polyfill` are both dependencies that should be added explicitly.

**Comment by levithomason** [comment]
Hm, checking.

**Comment by nfischer** [comment]
Thanks! I just used `nvm use v4` and then used my [github-review](https://github.com/nfischer/github-review) script to download a fresh copy of the PR.

**Comment by levithomason** [comment]
Cool, no worries.  You are right about the separate packages, but since we have `babel-cli`, we get those deps as well:

```shell
› npm ls babel-register
shx@0.0.1 /Users/levithomason/src/shx
`-- babel-cli@6.7.7
  `-- babel-register@6.7.2 

› npm ls babel-polyfill
shx@0.0.1 /Users/levithomason/src/shx
`-- babel-cli@6.7.7
  `-- babel-polyfill@6.7.4 
```

**Comment by nfischer** [comment]
After running `npm install` on Linux:

```bash
npm WARN package.json shx@0.0.1 No bin file found at lib/cli.js
npm WARN optional dep failed, continuing fsevents@1.0.11
npm WARN deprecated jade@0.26.3: Jade has been renamed to pug, please install the latest version of pug instead of jade
npm WARN deprecated graceful-fs@2.0.3: graceful-fs version 3 and before will fail on newer node releases. Please update to graceful-fs@^4.0.0 as soon as possible.
...
$ npm test # this command fails, as before
$ npm ls babel-register
shx@0.0.1 /home/nate/PR_31/node_modules/shx
├─┬ babel-cli@6.7.7
│ └── babel-register@6.7.2
└─┬ babel-preset-es2015@6.6.0
  └─┬ babel-plugin-transform-regenerator@6.6.5
    └─┬ babel-core@6.7.7
      └── babel-register@6.7.2

$ npm ls babel-polyfill
shx@0.0.1 /home/nate/PR_31/node_modules/shx
└─┬ babel-cli@6.7.7
  └── babel-polyfill@6.7.4
```

@levithomason Can you successfully re-clone, install deps, and test, all using node v4.2.4? If so, I think this may be an OS difference between Linux and OS X.

**Comment by nfischer** [comment]
Yeah, retried it again. For me, it all works if those packages are direct dependencies. Please let me know if I'm doing something wrong or if you're seeing something else.

**Comment by levithomason** [comment]
Confirmed and fixed.  npm v3 (node 5) installs modules in a flat tree, which allows requiring deps of other deps.  npm v2 (node 4) uses nested node_modules so they cannot be required.

Good catch and thanks for testing, I've added the deps explicitly.

**Comment by nfischer** [comment]
LGTM. Passes for both node v4 and v5.

I'll merge this in and rebase the other PR to see if things are working.

File: test/setup.js (hunk 3 above)
Comment (ariporad) [inline test/setup.js]: 👍

**Pull request merged**